File: sqlite-android-compiler/src/main/java/com/alecstrong/sqlite/android/TableGenerator.java
package com.alecstrong.sqlite.android;

import com.alecstrong.sqlite.android.model.Column;
import com.alecstrong.sqlite.android.model.ColumnConstraint;
import com.alecstrong.sqlite.android.model.SqlElement;
import com.alecstrong.sqlite.android.model.SqlStmt;
import com.alecstrong.sqlite.android.model.SqlStmt.Replacement;
import com.alecstrong.sqlite.android.model.Table;
import com.google.common.base.Joiner;
import java.io.File;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import static com.alecstrong.sqlite.android.SqliteCompiler.getFileExtension;

public abstract class TableGenerator<
    OriginatingType,
    SqliteStatementType extends OriginatingType,
    TableType extends OriginatingType,
    ColumnType extends OriginatingType,
    ConstraintType extends OriginatingType> extends SqlElement<OriginatingType> {
  private static final String CREATE_KEY_VALUE_TABLE = "\n"
      + "CREATE TABLE %s (\n"
      + "  " + SqliteCompiler.KEY_VALUE_KEY_COLUMN + " TEXT NOT NULL PRIMARY KEY,\n"
      + "  " + SqliteCompiler.KEY_VALUE_VALUE_COLUMN + " BLOB\n"
      + ");";
  private static final String CREATE_TABLE_IDENTIFIER = "createTable";

  public static final String outputDirectory = "generated/source/sqlite";

  private final Table<OriginatingType> table;
  private final List<SqlStmt<OriginatingType>> sqliteStatements;
  private final String interfaceName;
  private final String packageName;
  private final String projectPath;

  public TableGenerator(OriginatingType rootElement, String packageName, String sqliteFileName,
      String projectPath) {
    super(rootElement);
    this.packageName = packageName;
    this.projectPath = projectPath;
    this.interfaceName = sqliteFileName.endsWith(getFileExtension()) //
        ? sqliteFileName.substring(0, sqliteFileName.length() - (getFileExtension().length() + 1))
        : sqliteFileName;
    this.sqliteStatements = new ArrayList<SqlStmt<OriginatingType>>();

    Table<OriginatingType> table = null;
    try {
      TableType tableElement = tableElement(rootElement);
      if (tableElement != null) {
        List<Replacement> replacements = new ArrayList<Replacement>();
        table = tableFor(tableElement, packageName, interfaceName, replacements);
        if (table.isKeyValue()) {
          sqliteStatements.add(new SqlStmt<OriginatingType>(CREATE_TABLE_IDENTIFIER,
              String.format(CREATE_KEY_VALUE_TABLE, tableName(tableElement)), 0,
              Collections.<Replacement>emptyList(), tableElement));
        } else {
          sqliteStatements.add(new SqlStmt<OriginatingType>( //
              CREATE_TABLE_IDENTIFIER, text(tableElement), startOffset(tableElement), replacements,
              tableElement));
        }
      }

      for (SqliteStatementType sqlStatementElement : sqlStatementElements(rootElement)) {
        List<Replacement> replacements = new ArrayList<Replacement>();
        if (identifier(sqlStatementElement) != null) {
          sqliteStatements.add(sqliteStatementFor(sqlStatementElement, replacements));
        }
      }
    } catch (ArrayIndexOutOfBoundsException e) {
      // Do nothing, just an easy way to catch a lot of situations where sql is incomplete.
      table = null;
    }
    this.table = table;
  }

  protected abstract Iterable<SqliteStatementType> sqlStatementElements(
      OriginatingType originatingElement);

  protected abstract TableType tableElement(OriginatingType sqlStatementElement);

  protected abstract String identifier(SqliteStatementType sqlStatementElement);

  protected abstract Iterable<ColumnType> columnElements(TableType tableElement);

  protected abstract String tableName(TableType tableElement);

  protected abstract boolean isKeyValue(TableType tableElement);

  protected abstract String columnName(ColumnType columnElement);

  protected abstract String classLiteral(ColumnType columnElement);

  protected abstract String typeName(ColumnType columnElement);

  protected abstract Replacement replacementFor(ColumnType columnElement, Column.Type type);

  protected abstract Iterable<ConstraintType> constraintElements(ColumnType columnElement);

  protected abstract ColumnConstraint<OriginatingType> constraintFor(
      ConstraintType constraintElement, List<Replacement> replacements);

  protected abstract String text(OriginatingType sqliteStatementElement);

  protected abstract int startOffset(OriginatingType sqliteStatementElement);

  private Table<OriginatingType> tableFor(TableType tableElement, String packageName,
      String fileName, List<Replacement> replacements) {
    Table<OriginatingType> table =
        new Table<OriginatingType>(packageName, fileName, tableName(tableElement), tableElement,
            isKeyValue(tableElement));

    for (ColumnType columnElement : columnElements(tableElement)) {
      table.addColumn(columnFor(columnElement, replacements));
    }
    return table;
  }

  private Column<OriginatingType> columnFor(ColumnType columnElement,
      List<Replacement> replacements) {
    String columnName = columnName(columnElement);
    Column<OriginatingType> result;
    Column.Type type = Column.Type.valueOf(typeName(columnElement));
    replacements.add(replacementFor(columnElement, type));
    if (classLiteral(columnElement) != null) {
      result =
          new Column<OriginatingType>(columnName, type, classLiteral(columnElement), columnElement);
    } else {
      result = new Column<OriginatingType>(columnName, type, columnElement);
    }

    for (ConstraintType constraintElement : constraintElements(columnElement)) {
      ColumnConstraint<OriginatingType> constraint = constraintFor(constraintElement, replacements);
      if (constraint != null) result.addConstraint(constraint);
    }
    return result;
  }

  private SqlStmt<OriginatingType> sqliteStatementFor(SqliteStatementType sqliteStatementElement,
      List<Replacement> replacements) {
    return new SqlStmt<OriginatingType>(
        identifier(sqliteStatementElement),
        text(sqliteStatementElement),
        startOffset(sqliteStatementElement),
        replacements,
        sqliteStatementElement);
  }

  public Table<OriginatingType> table() {
    return table;
  }

  public File getOutputDirectory() {
    return new File(projectPath + "build/" + outputDirectory);
  }

  public String fileName() {
    return interfaceName() + ".java";
  }

  /**
   * @return the package directory structure for the generated file (format: 'com/sample/package')
   */
  public String packageDirectory() {
    return Joiner.on('/').join(packageName().split("\\."));
  }

  String packageName() {
    return packageName;
  }

  File getFileDirectory() {
    return new File(getOutputDirectory(), packageDirectory());
  }

  String interfaceName() {
    return interfaceName + "Model";
  }

  List<SqlStmt<OriginatingType>> sqliteStatements() {
    return sqliteStatements;
  }
}

File: sqldelight-compiler/src/main/kotlin/com/squareup/sqldelight/TableGenerator.kt
/*
 * Copyright (C) 2016 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.squareup.sqldelight

import com.squareup.sqldelight.model.Column
import com.squareup.sqldelight.model.Column.Type
import com.squareup.sqldelight.model.ColumnConstraint
import com.squareup.sqldelight.model.SqlElement
import com.squareup.sqldelight.model.SqlStmt
import com.squareup.sqldelight.model.SqlStmt.Replacement
import com.squareup.sqldelight.model.Table
import org.antlr.v4.runtime.ParserRuleContext
import java.io.File
import java.io.File.separatorChar
import java.util.ArrayList

class TableGenerator
constructor(
    rootElement: SqliteParser.ParseContext,
    relativeFile: String,
    private val projectPath: String
) : SqlElement(rootElement) {
  private val fileName = relativeFile.split(separatorChar).last()
  private val originalFileName =
      if (fileName.endsWith(SqliteCompiler.FILE_EXTENSION)) //
        fileName.substring(0, fileName.length - (SqliteCompiler.FILE_EXTENSION.length + 1))
      else fileName

  internal val table: Table?
  internal val sqliteStatements = ArrayList<SqlStmt>()
  internal val packageName = relativeFile.split(separatorChar).dropLast(1).joinToString(".")

  val generatedFileName = originalFileName + "Model"
  val outputDirectory = File(File(projectPath, "build"), SqliteCompiler.OUTPUT_DIRECTORY)
  val packageDirectory = packageName.replace('.', separatorChar)
  val fileDirectory = File(outputDirectory, packageDirectory)

  init {
    if (packageName.isEmpty()) {
      throw SqlitePluginException(rootElement, ".sq cannot be children of the sqldelight " +
          "container. Place them in their own package under sqldelight.")
    }
    var table: Table? = null
    try {
      val tableElement = rootElement.sql_stmt_list().create_table_stmt()
      if (tableElement != null) {
        val replacements = ArrayList<Replacement>()
        table = tableFor(tableElement, packageName, originalFileName, replacements)
        sqliteStatements.add(SqlStmt("create_table", text(tableElement),
            startOffset(tableElement), replacements, tableElement))
      }

      for (sqlStatementElement in rootElement.sql_stmt_list().sql_stmt()) {
        sqliteStatements.add(sqliteStatementFor(sqlStatementElement, arrayListOf()))
      }
    } catch (e: ArrayIndexOutOfBoundsException) {
      // Do nothing, just an easy way to catch a lot of situations where sql is incomplete.
      table = null
    }

    this.table = table
  }

  fun startOffset(sqliteStatementElement: ParserRuleContext) =
      when (sqliteStatementElement) {
        is SqliteParser.Create_table_stmtContext -> sqliteStatementElement.start.startIndex
        else -> (sqliteStatementElement.getChild(
            sqliteStatementElement.childCount - 1) as ParserRuleContext).start.startIndex
      }

  fun text(sqliteStatementElement: ParserRuleContext) = when (sqliteStatementElement) {
    is SqliteParser.Sql_stmtContext -> sqliteStatementElement.statementTextWithWhitespace()
    else -> sqliteStatementElement.textWithWhitespace()
  }

  private fun tableFor(tableElement: SqliteParser.Create_table_stmtContext, packageName: String,
      fileName: String, replacements: MutableList<Replacement>): Table {
    val table = Table(packageName, fileName, tableElement.table_name().text, tableElement)
    tableElement.column_def().forEach { table.columns.add(columnFor(it, replacements)) }
    return table
  }

  private fun columnFor(columnElement: SqliteParser.Column_defContext,
      replacements: MutableList<Replacement>): Column {
    val columnName = columnElement.column_name().text
    val type = Type.valueOf(typeName(columnElement))
    val constraints = columnElement.column_constraint()
        .map({ constraintFor(it, replacements) })
        .filterNotNull()
    val classLiteral = columnElement.type_name().sqlite_class_name()?.STRING_LITERAL()?.text?.filter { it != '\'' }
    val result = when (classLiteral) {
      null -> Column(columnName, type, constraints, originatingElement = columnElement)
      else -> Column(columnName, type, constraints,  classLiteral, columnElement)
    }

    replacements.add(Replacement(columnElement.type_name().start.startIndex,
            columnElement.type_name().stop.stopIndex + 1, type.replacement))

    return result
  }

  private fun sqliteStatementFor(sqliteStatementElement: SqliteParser.Sql_stmtContext,
      replacements: List<Replacement>): SqlStmt =
      SqlStmt(sqliteStatementElement.sql_stmt_name().text, text(sqliteStatementElement),
          startOffset(sqliteStatementElement), replacements, sqliteStatementElement)

  private fun constraintFor(constraintElement: SqliteParser.Column_constraintContext,
      replacements: List<Replacement>): ColumnConstraint? =
      when {
        constraintElement.K_NOT() != null -> ColumnConstraint.NotNullConstraint(constraintElement);
        else -> null
      }

  private fun typeName(columnElement: SqliteParser.Column_defContext) =
      when {
        columnElement.type_name().sqlite_class_name() != null ->
          columnElement.type_name().sqlite_class_name().getChild(0).text
        else -> columnElement.type_name().sqlite_type_name().text
      }
}

fun String.relativePath(originatingElement: ParserRuleContext): String {
  val parts = split(separatorChar)
  for (i in 2..parts.size) {
    if (parts[i - 2] == "src" && parts[i] == "sqldelight") {
      return parts.subList(i + 1, parts.size).joinToString(separatorChar.toString())
    }
  }
  throw SqlitePluginException(originatingElement,
      "Files must be organized like src/main/sqldelight/...")
}
